dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -7,3263 +7,6479 @@
-            if (!skip_struct_member_at_index(i, num_skipped)) {
-              break;
-            }
-            total_skipped += num_skipped;
-            if (total_skipped == index + 1) {
-              DDS::DynamicTypeMember_var member;
-              if (base_type->get_member_by_index(member, i) == DDS::RETCODE_OK) {
-                id = member->get_id();
-              }
-              break;
-            }
-          }
-          return id;
-        }
-      } else { // Mutable
-        ACE_CDR::ULong member_id;
-        size_t member_size;
-        bool must_understand;
-        bool good = true;
-        for (unsigned i = 0; i < index; ++i) {
-          if (!strm_.read_parameter_id(member_id, member_size, must_understand)) {
-            good = false;
-            break;
-          }
-
-          DDS::DynamicTypeMember_var dtm;
-          if (base_type->get_member(dtm, member_id) != DDS::RETCODE_OK) {
-            good = false;
-            break;
-          }
-          DDS::MemberDescriptor_var descriptor;
-          if (dtm->get_descriptor(descriptor) != DDS::RETCODE_OK) {
-            good = false;
-            break;
-          }
-          const DDS::DynamicType_ptr mt = descriptor->type();
-          if (!mt) {
-            good = false;
-            break;
-          }
-          const DDS::DynamicType_var member = get_base_type(mt);
-          if (member->get_kind() == TK_SEQUENCE) {
-            if (!skip_sequence_member(member)) {
-              good = false;
-              break;
-            }
-          } else if (!strm_.skip(member_size)) {
-            good = false;
-            break;
-          }
-        }
+  if (member_tk != MemberTypeKind && member_tk != enum_or_bitmask) {
+    return false;
+  }
 
-        if (!good || !strm_.read_parameter_id(member_id, member_size, must_understand)) {
-          member_id = MEMBER_ID_INVALID;
-        }
-        return member_id;
-      }
+  if (member_tk == enum_or_bitmask) {
+    DDS::TypeDescriptor_var member_td;
+    if (member_type->get_descriptor(member_td) != DDS::RETCODE_OK) {
+      return false;
     }
-  case TK_UNION:
-    {
-      if (index == 0) {
-        return DISCRIMINATOR_ID;
-      }
-
-      // Get the Id of the selected member.
-      DDS::MemberDescriptor_var selected_md = get_union_selected_member();
-      if (!selected_md) {
-        return MEMBER_ID_INVALID;
-      }
-
-      return selected_md->id();
+    const CORBA::ULong bit_bound = member_td->bound()[0];
+    if (bit_bound < lower || bit_bound > upper) {
+      return false;
     }
   }
 
-  if (DCPS::DCPS_debug_level >= 1) {
-    ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) DynamicDataImpl::get_member_id_at_index -")
-               ACE_TEXT(" Calling on an unexpected type %C\n"), typekind_to_string(tk)));
-  }
-  return MEMBER_ID_INVALID;
+  return insert_single(id, value);
 }
 
-ACE_CDR::ULong DynamicDataImpl::get_item_count()
+bool DynamicDataImpl::is_discriminator_type(TypeKind tk) const
```

Comment:
Should this be named something like is_valid_discriminator_type or is_possible_discriminiator_type?  It can be a `static` member since it doesn't use `this`.  That would make it clear that it's not checking if `tk` is the type of this object's discriminator.